ansible/plugins/connection/multi_passwd_ssh.py

```diff
@@ -96,23 +112,15 @@ def wrapped(self, *args, **kwargs):
             ipv4_addr_unavailable = (CONNECTION_TIMEOUT_ERR_FLAG1 in e.message) or \
                                     (CONNECTION_TIMEOUT_ERR_FLAG2 in e.message)
 
-            try_ipv6_addr = (type(e) != AnsibleAuthenticationFailure) and ipv4_addr_unavailable and hostv6
-            if try_ipv6_addr:
-                # This is a retry, so the fd/pipe for sshpass is closed, and we need a new one
-                self.sshpass_pipe = os.pipe()
-                self._play_context.remote_addr = hostv6
-                # args sample:
-                # ( [b'sshpass', b'-d18', b'ssh', b'-o', b'ControlMaster=auto', b'-o', b'ControlPersist=120s', b'-o', b'UserKnownHostsFile=/dev/null', b'-o', b'StrictHostKeyChecking=no', b'-o', b'StrictHostKeyChecking=no', b'-o', b'User="admin"', b'-o', b'ConnectTimeout=60', b'-o', b'ControlPath="/home/user/.ansible/cp/376bdcc730"', 'fc00:1234:5678:abcd::2', b'/bin/sh -c \'echo PLATFORM; uname; echo FOUND; command -v \'"\'"\'python3.10\'"\'"\'; command -v \'"\'"\'python3.9\'"\'"\'; command -v \'"\'"\'python3.8\'"\'"\'; command -v \'"\'"\'python3.7\'"\'"\'; command -v \'"\'"\'python3.6\'"\'"\'; command -v \'"\'"\'python3.5\'"\'"\'; command -v \'"\'"\'/usr/bin/python3\'"\'"\'; command -v \'"\'"\'/usr/libexec/platform-python\'"\'"\'; command -v \'"\'"\'python2.7\'"\'"\'; command -v \'"\'"\'/usr/bin/python\'"\'"\'; command -v \'"\'"\'python\'"\'"\'; echo ENDFOUND && sleep 0\''], None) # noqa: E501
-                # args[0] are the parameters of ssh connection
-                ssh_args = args[0]
-                # Change the IPv4 host in the ssh_args to IPv6
-                for idx in range(len(ssh_args)):
-                    if type(ssh_args[idx]) == bytes and ssh_args[idx].decode() == self.host:
-                        ssh_args[idx] = hostv6
-                self.host = hostv6
-                self.set_option("host", hostv6)
+            try_ipv6_addr = orig_host != hostv6 and (type(e) != AnsibleAuthenticationFailure) and \
+                ipv4_addr_unavailable and hostv6
+            if not try_ipv6_addr:
+                raise e
+            self._change_host(hostv6, *args)
+            try:
                 return _conn_with_multi_pwd(self, *args, **kwargs)
-            else:
+            except AnsibleConnectionFailure as e:
```

Comment:
Is it better to always transfer back to IPv4 no matter what kind of error it is?

Reply:
If it is authentication error, then it means ipv6 is working. When ipv4 fails, current code also only tries ipv6 when there is connection error.

Reply:
I reckon the purpose of your change is we want to convert back to IPv4 anyway?
Even though it's AnsibleConnectionFailure who target the logic to this line.
As you said, we could hit authentication error during the retry on IPv6.
In this situation, should we roll back to IPv4 also?

It's just a reminder, you're blocked by the mgmt-ip changing behavior, you made the call whether always-fall-back is needed.

Reply:
I think the current logic is, try ipv4 first, if authentication error, raise the error, if connection error, try ipv6, if authentication error, raise the error, if connection error, change back to ipv4 so ipv4 will be tried in the future